### DISCUSSION TITLE
Sleigh: instruction with mixed MSB and LSB parameters

### DISCUSSION THREAD

**Discussion by inode-** [NationalSecurityAgency/ghidra#5141] [Q&A]
I'm working on a BE architecture (similar to openrisc) which has in some instructions a mix of MSB and LSB paramters.

For example has the instruction `movi D,G` where D is 5 bits and G is 4 bits. The problem is that D is in MSB and so it's interpreted correctly, but G is in LSB so in interpreted in a wrong way.

```
define token b0(16)
b_16_op_0_3 = (12,15)
b_16_op_4_5 = (10,11)
b_16_op_11_11 = (4,4)
b_16_v_A_11_15 = (0,4)
b_16_v_D_6_10 = (5,9)
b_16_v_G_12_15 = (0,3) 
b_16_v_T_6_15 = (0,9)
;

:bt.movi b_16_v_D_6_10,b_16_v_G_12_15 is b_16_op_0_3=0x0 & b_16_op_4_5=0b00 & b_16_v_D_6_10 & b_16_op_11_11=0b0 & b_16_v_G_12_15 { } 
```
I have tried to find a way to specify with the SLEIGH language that a specific parameter is LSB or convert it to LSB, but I didn't find a way to do it.

Any idea?

Thank you

**Comment by GhidorahRex** [comment]
You can attach values to a token field. Since it's only 4 bits wide, it should be simple enough to enumerate all 16 possibilities. Check out the 68000.sinc file and look at how we handle the "quick" field.

**Comment by inode-** [comment]
Unfortunately some of the instructions manage 32 bit values... so the "quick" approach is not feasible.

**Comment by inode-** [comment]
Rebuild the variable in the reverse order it's an option:

```
a = (0,0)
b = (1,1)
c = (2,2)
d = (3,3)
;

:bt.movi b_16_v_D_6_10,tmp is b_16_op_0_3=0x0 & b_16_op_4_5=0b00 & b_16_v_D_6_10 & b_16_op_11_11=0b0 & b_16_v_G_12_15 & a & b & c & d [ tmp = ((a << 3) | (b << 2) | (c << 1) | d) ; ] { }
```
This code for example create a new variable reversing LSB/MSB but it will create a lot of code inside the sleigh. Any other option?

**Comment by GhidorahRex** [comment]
If you have a 32-bit value that is reversing endianness then your solution is probably the fastest. I agree, it's a little messy, but it should suffice. However, I would create a subconstructor for it, e.g.:

```
a = (0,7)
b = (8,15)
c = (16,23)
d = (24,31)

attach values [a b c d] [
 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15
16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31
32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63
64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79
80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95
96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111
112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127
];

lsb32: value is a & b & c & d [value = (a << 24) | (b << 16) | (c << 8) | d); ] { local tmp:4 = value; export tmp; }
```

edit: you would want to reverse the numbers in the attach now that I think about it.

**Comment by inode-** [comment]
Just to leave a trace for all the people who have the same problem. LSB vs MSB is not simply the opposite of the numbers. Find the correct numbers below.

```
# Values of 2 bits
attach values [ a ] [0 2 1 3 ];

# Values of 3 bits
attach values [ b ] [0 4 2 6 1 5 3 7 ];

# Values of 4 bits
attach values [ c ] [0 8 4 12 2 10 6 14 1 9 5 13 3 11 7 15 ];

# Values of 5 bits
attach values [ d ] [0 16 8 24 4 20 12 28 2 18 10 26 6 22 14 30 1 17 9 25 5 21 13 29 3 19 11 27 7 23 15 31 ];

# Values of 6 bits
attach values [ e ] [0 32 16 48 8 40 24 56 4 36 20 52 12 44 28 60 2 34 18 50 10 42 26 58 6 38 22 54 14 46 30 62 1 33 17 49 9 41 25 57 5 37 21 53 13 45 29 61 3 35 19 51 11 43 27 59 7 39 23 55 15 47 31 63 ];

# Values of 7 bits
attach values [ f ] [0 64 32 96 16 80 48 112 8 72 40 104 24 88 56 120 4 68 36 100 20 84 52 116 12 76 44 108 28 92 60 124 2 66 34 98 18 82 50 114 10 74 42 106 26 90 58 122 6 70 38 102 22 86 54 118 14 78 46 110 30 94 62 126 1 65 33 97 17 81 49 113 9 73 41 105 25 89 57 121 5 69 37 101 21 85 53 117 13 77 45 109 29 93 61 125 3 67 35 99 19 83 51 115 11 75 43 107 27 91 59 123 7 71 39 103 23 87 55 119 15 79 47 111 31 95 63 127 ];

# Values of 8 bits
attach values [ g ] [0 128 64 192 32 160 96 224 16 144 80 208 48 176 112 240 8 136 72 200 40 168 104 232 24 152 88 216 56 184 120 248 4 132 68 196 36 164 100 228 20 148 84 212 52 180 116 244 12 140 76 204 44 172 108 236 28 156 92 220 60 188 124 252 2 130 66 194 34 162 98 226 18 146 82 210 50 178 114 242 10 138 74 202 42 170 106 234 26 154 90 218 58 186 122 250 6 134 70 198 38 166 102 230 22 150 86 214 54 182 118 246 14 142 78 206 46 174 110 238 30 158 94 222 62 190 126 254 1 129 65 193 33 161 97 225 17 145 81 209 49 177 113 241 9 137 73 201 41 169 105 233 25 153 89 217 57 185 121 249 5 133 69 197 37 165 101 229 21 149 85 213 53 181 117 245 13 141 77 205 45 173 109 237 29 157 93 221 61 189 125 253 3 131 67 195 35 163 99 227 19 147 83 211 51 179 115 243 11 139 75 203 43 171 107 235 27 155 91 219 59 187 123 251 7 135 71 199 39 167 103 231 23 151 87 215 55 183 119 247 15 143 79 207 47 175 111 239 31 159 95 223 63 191 127 255 ];
```

However, it would be very useful to have these conversions in the sleigh language so that they can also be associated with other information such as the "signed" keyword.

**Comment by inode-** [comment]
I have run further tests but it seems that the instructions executed inside the brackets take the original value and not the one specified in the "attached values". Also looking in the (ghidra documentation)[https://ghidra.re/courses/languages/html/sleigh_constructors.html#sleigh_disassembly_actions]  I get: 

`For certain kinds of instructions, there is a need to calculate values that depend on the specific bits of the instruction, but which cannot be obtained as an integer interpretation of a field or by building with an attach values statement.
`

For example the code:

```
define token b3(48)
b_48_v_A_11_15 = (32,36)
b_48_v_g_40_47 = (0,7)
b_48_v_g_16_23 = (24,31)
b_48_v_g_24_31 = (16,23)
b_48_v_g_32_39 = (8,15)
;

attach values [ b_48_v_g_40_47  b_48_v_g_16_23 b_48_v_g_24_31  b_48_v_g_32_39 ] [0 128 64 192 32 160 96 224 16 144 80 208 48 176 112 240 8 136 72 200 40 168 104 232 24 152 88 216 56 184 120 248 4 132 68 196 36 164 100 228 20 148 84 212 52 180 116 244 12 140 76 204 44 172 108 236 28 156 92 220 60 188 124 252 2 130 66 194 34 162 98 226 18 146 82 210 50 178 114 242 10 138 74 202 42 170 106 234 26 154 90 218 58 186 122 250 6 134 70 198 38 166 102 230 22 150 86 214 54 182 118 246 14 142 78 206 46 174 110 238 30 158 94 222 62 190 126 254 1 129 65 193 33 161 97 225 17 145 81 209 49 177 113 241 9 137 73 201 41 169 105 233 25 153 89 217 57 185 121 249 5 133 69 197 37 165 101 229 21 149 85 213 53 181 117 245 13 141 77 205 45 173 109 237 29 157 93 221 61 189 125 253 3 131 67 195 35 163 99 227 19 147 83 211 51 179 115 243 11 139 75 203 43 171 107 235 27 155 91 219 59 187 123 251 7 135 71 199 39 167 103 231 23 151 87 215 55 183 119 247 15 143 79 207 47 175 111 239 31 159 95 223 63 191 127 255 ];

:bw.sfgtui b_48_v_A_11_15,b_48_v_g_16_47,b_48_v_g_40_47,b_48_v_g_32_39,b_48_v_g_24_31,b_48_v_g_16_23 is b_48_op_0_3=0x9 & b_48_op_4_9=0b111011 & b_48_v_A_11_15  & b_48_v_g_40_47  & b_48_v_g_32_39  & b_48_v_g_24_31  & b_48_v_g_16_23 [  b_48_v_g_16_47 = (  ( b_48_v_g_40_47 << 24 ) |  ( b_48_v_g_32_39 << 16 ) |  ( b_48_v_g_24_31 << 8 ) |  b_48_v_g_16_23  ); ]
{

}
```

The output expected should be the same in the 4 different parameter and the first one, but it differ, so the documentation is correct and this way is not usable in my case.

Output: `bw.sfgtui  r3,0xe0ffff,0x0,0x7,0xff,0xff`

@GhidorahRex any other way?

**Comment by GhidorahRex** [comment]
Do you have an instruction manual for this language?

**Comment by inode-** [comment]
Unfortunatelly no instruction manual is available.

In this case the objectdump (taken from the SDK) of "9e c3 ff ff e0 00" is: "b.sfgtui   r3,0x7ffff". First parameter is MSB, second id LSB.

At PCODE level I can implement transaction to LSB on the second parameter, but at disassembler level it seems to be not possible.

The only way that I see is to modify the sleigh language to implement a field keywork (like signed, dec, hex) to convert the data to LSB. I was trying to understand correctly how to modify the code to implement this enachment.

**Comment by mumbel** [comment]
https://github.com/GravisZro/BA2-toolchain

Gcc/Binutils source at least

**Comment by inode-** [comment]
I try to make a recap from everything I have learned hoping it will be useful to someone else.

The processor I'm working on is a big endiand and interprets some parameters as LSB (least significant bit) and others as MSB (most significant bit).

One example is `b.movi r9, 0xC` which in bytes is `0x01 0x23`

Translated in bits:

```
                          0x01     0x23
                        00000001 00100011
instruction selection   XXXXXX      X
param1                        XX XXX
param2                               XXXX
```

param1 is the destination registry, in this case 01001 (9 in decimal) which will be R9 (correct).
param2 is the value, if parsed like Ghidra does the value will be 0011 (3 in decimal) which is incorrect because this parameter need to be parsed in LSB.

I have not been able to identify an easy way to do this with the sleigh language (I think it is a limitation of the language), I have been forced to use the "Disassembly Actions Section" to reverse all LSB values bit by bit. The mechanism works, but on large values a long and complex string is generated.

```
define token b0(16)
b_16_op_4_4 = (4,4)
b_16_op_10_11 = (10,11)
b_16_op_12_15 = (12,15)
b_16_v_A_0_4 = (0,4)
b_16_v_D_5_9 = (5,9)
b_16_b0 = (0,0)
b_16_b1 = (1,1)
b_16_b2 = (2,2)
b_16_b3 = (3,3)
b_16_b4 = (4,4)
b_16_b5 = (5,5)
b_16_b6 = (6,6)
b_16_b7 = (7,7)
b_16_b8 = (8,8)
b_16_b9 = (9,9)
b_16_b10 = (10,10)
b_16_b11 = (11,11)
b_16_b12 = (12,12)
b_16_b13 = (13,13)
b_16_b14 = (14,14)
b_16_b15 = (15,15)
;

attach variables [b_16_v_D_5_9 ] [ r0 r1 r2 r3 r4 r5 r6 r7 r8 r9 r10 r11 r12 r13 r14 r15 r16 r17 r18 r19 r20 r21 r22 r23 r24 r25 r26 r27 r28 r29 r30 r31 ];

:bt.movi b_16_v_D_5_9,b_16_v_G_0_3 is b_16_op_12_15=0x0 & b_16_op_10_11=0b00 & b_16_v_D_5_9  & b_16_op_4_4=0b0 & b_16_b0  & b_16_b1  & b_16_b2  & b_16_b3 [  b_16_v_G_0_3 = (  ( b_16_b0 << 3 ) |  ( b_16_b1 << 2 ) |  ( b_16_b2 << 1 ) |  (b_16_b3)  ); ] 
{

}
```

In this case the new variable `b_16_v_G_0_3` generated by the single bits correctly interprets param2 as 1100 (12 in decimal).

_"attach values" cannot be used to speedup the process because are not accepted into the "disassembly actions section"_
